msal/src/test/java/com/microsoft/identity/client/e2e/tests/network/nativeauth/SignInEmailOTPTest.kt

```diff
@@ -0,0 +1,96 @@
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+
+package com.microsoft.identity.client.e2e.tests.network.nativeauth
+
+import com.microsoft.identity.client.e2e.utils.assertState
+import com.microsoft.identity.internal.testutils.nativeauth.ConfigType
+import com.microsoft.identity.internal.testutils.nativeauth.api.TemporaryEmailService
+import com.microsoft.identity.nativeauth.statemachine.errors.SignInError
+import com.microsoft.identity.nativeauth.statemachine.errors.SubmitCodeError
+import com.microsoft.identity.nativeauth.statemachine.results.SignInResult
+import com.microsoft.identity.nativeauth.statemachine.results.SignUpResult
+import kotlinx.coroutines.runBlocking
+import kotlinx.coroutines.test.runTest
+import org.junit.Assert
+import org.junit.Test
+
+class SignInEmailOTPTest : NativeAuthPublicClientApplicationAbstractTest() {
+
+    private val tempEmailApi = TemporaryEmailService()
+
+    override val configType = ConfigType.SIGN_IN_OTP
+
+    /**
+     * Use valid email and OTP to get token and sign in.
+     * (hero scenario 6, use case 2.2.1, Test case 30)
+     */
+    @Test
+    fun testSuccess() {
+        retryOperation {
+            runBlocking { // Running with runBlocking to avoid default 10 second execution timeout.
+                val user = config.email
+                val signInResult = application.signIn(user)
+                assertState<SignInResult.CodeRequired>(signInResult)
+                val otp = tempEmailApi.retrieveCodeFromInbox(user)
+                val submitCodeResult = (signInResult as SignInResult.CodeRequired).nextState.submitCode(otp)
+                assertState<SignInResult.Complete>(submitCodeResult)
+            }
+        }
+    }
+
+    /**
+     * Use invalid email address to receive a "user not found" error.
+     * (use case 2.2.2, Test case 31)
+     */
+    @Test
+    fun testErrorIsUserNotFound() = runTest {
+        val user = config.email
+        // Turn correct username into an incorrect one
+        val invalidUser = user + "x"
+        val signInResult = application.signIn(invalidUser)
+        Assert.assertTrue(signInResult is SignInError)
+        Assert.assertTrue((signInResult as SignInError).isUserNotFound())
+    }
+
+    /**
+     * Use valid email address, but invalid OTP to receive "invalid code" error.
+     * (use case 2.2.7, Test case 35)
+     */
+    @Test
+    fun testErrorIsInvalidCode() {
+        retryOperation {
+            runBlocking {// Running with runBlocking to avoid default 10 second execution timeout.
+                val user = config.email
+                val signInResult = application.signIn(user)
+                assertState<SignInResult.CodeRequired>(signInResult)
+                val otp = tempEmailApi.retrieveCodeFromInbox(user)
```

Comment:
Since we want to sent an invalid otp why not avoid retrieving the correct one from the email? is the intention to check that the otp has been sent successfully?

Reply:
In theory, setting the "invalid" value to a predefined, static one could coincidentally be the correct OTP. The only way to guarantee an incorrect one is by taking the correct one and editing it to something else. But the likelihood of that is very small, so I agree this is a good optimisation to make.

Reply:
We can set the predefined invalid OTP to be a short one (if I'm not mistaken the OTPs are usually 7 or 8 digits) or give it a really bogus value like 00000 to avoid this edge case